reps/2023-02-03-node-affinity-feature-enhancements.md

```diff
@@ -0,0 +1,399 @@
+}
+```
+
+### If the Match Expression Cannot be satisfied
+If the matching expression cannot be satisfied, The actor/task will be add to the pending queue. Util the matching expression all be statisfied.
+
+1. resources are enough, but node_id of affinity don't exist. -> throw exception, compatible with current behavior 
+
+2. resources are enough, but labels affinity  cannot be satisfied? -> Hang & report schedule failed event and detail unstaisfed reason to exposed to users
+
+3. labels affinity can be satisfied, but resources are not enough? -> Hang & report schedule failed event and detail unstaisfed reason to exposed to users
+
+4. both labelsaffinity and resources are not satisfied? -> Hang & report schedule failed event and detail unstaisfed reason to exposed to users
+
+
+### how other system achieve the same goal?
+1、K8s
+This solution is to learn the PodAffinity/NodeAffinity features of K8s。
+https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity
+
+Scheduling Policy | Label Owner | Operator
+-- | -- | --
+nodeAffinity | NODE | In, NotIn, Exists, DoesNotExist, Gt, Lt
+podAffinity | POD | In, NotIn, Exists, DoesNotExist
+podAntiAffinity | POD | In, NotIn, Exists, DoesNotExist
+
+
+### what's the alternative to achieve the same goal?
+** Option 3: putting Labels into the custom resource solution  **  
+```
+class ResourceRequest {
+    absl::flat_hash_map<ResourceID, FixedPoint> resources_;
+}
+
+Add Actor/Task/Node labels to resources_ as custom resources.
+eg:  
+{
+    "CPU": 16,
+    "memory": xx,
+    "custom_resources": xx,
+    "actor_labels_key@value": 1,
+    "task_labels_key@value": 1,
+    "node_labels_key@value": 1,
+}
+```
+If you put labels into custom_resources, you need to do the following adaptation:
+1. Compared with custom_resources, labels need to add a specific prefix to distinguish them from custom_resources.
+2. The key and value of Labels need to be concatenated with special characters (@).
+3. When using Labels to build a Labels index table, you need to parse the resources key.
+
+**DisAdvantages:**
+1. Labels unlike cpu resource these are numeric types. Compared with the above scheme. This will destroy the concept of coustrom resouce.
+2. Actor and Task are isolated by namespace. It is difficult to isolate through namespace if adding custom_resource.
+2. The Label index table of all nodes can be constructed from the ActorLabels information of each node. If you use Custom Resource, this requires parsing the resource_key and doing a lot of string splitting which will cost performance.
+3. If custom_resource happens to be the same as the spliced string of labels. Then it will affect the correctness of scheduling.
+
+### AutoScaler adaptation
+I understand that the adaptation of this part of AutoScaler should not be difficult. Now NodeAffinity is just one more scheduling strategy. Just follow the existing implementation of AutoScaler and adapt to the current strategy.
```

Comment:
This proposal is more advanced than the autoscaling for NodeAffinity.

NodeAffinity is only used for nodes that are already in the cluster, so I believe the only change there was to make sure that the autoscaler didn't mistakenly start additional nodes for a queued task that wouldn't actually be able to be scheduled there.

Since this proposal is for more flexible labels, we will need the autoscaler to be aware of which nodes it should add or remove from the cluster based on the label constraints of queued tasks. So this actually is something the proposal should cover. Specifically:
- what API changes do we need at the cluster config level and for the interface between the autoscaler and the GCS?
- what will the autoscaler policy look like?
- Handling scalability issues? (If an application uses too many different label policies, we'll have to report these in the task load)

Reply:
Yeah I don't think this is quite so simple. The autoscaler runs emulated bin-packing to determine what types of nodes it needs to add to the cluster. For example, suppose a user requests {"dc": "zone-1"}. Then, the autoscaler will need to run its bin packing routine with this request and node labels in mind, in order to determine the nodes to launch: https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/_private/resource_demand_scheduler.py

As you can see, the logic is algorithmically complex.

My assessment is that we probably need to refactor the autoscaler to unify this bin packing routine with the C++ implementation, to avoid having to write label affinity evaluation in both Python and C++.